src/ucx_van.h

```diff
@@ -441,195 +526,332 @@ class UCXVan : public Van {
+  {
+    UCXRequest *req = reinterpret_cast<UCXRequest*>(request);
+
+    CHECK_STATUS(status) << "TX request completed with " << ucs_status_string(status);
+
+    delete [] req->data.raw_meta;
+
+    UCX_REQUEST_FREE(req);
+  }
+
+  UCXEndpointsPool                                  ep_pool_;
+  ucp_context_h                                     context_;
+  ucp_worker_h                                      worker_;
+  ucp_listener_h                                    listener_;
+  int                                               src_dev_idx_;
+  Node                                              *my_node_;
+  UCXVan                                            *van_;
+  UCXRecvPool                                       *rx_pool_;
+};
+
+class UCXVan : public Van {
+ public:
+  UCXVan() {
+    setenv("UCX_USE_MT_MUTEX", "y", 0);
+    setenv("UCX_IB_NUM_PATHS", "2", 0);
+    setenv("UCX_SOCKADDR_CM_ENABLE", "y", 0);
+    setenv("UCX_RNDV_THRESH", "8k", 0);
+    short_send_thresh_ = GetEnv("BYTEPS_UCX_SHORT_THRESH", 4096);
+  }
+
+  ~UCXVan() {}
+
+  virtual std::string GetType() const {
+    return std::string("ucx");
+  }
+
+ protected:
+  void Start(int customer_id, bool standalone) override {
+    start_mu_.lock();
+    should_stop_ = false;
+
+    rx_pool_.reset(new UCXRecvPool());
+    start_mu_.unlock();
+
+    if (!standalone) Van::Start(customer_id, false);
+  }
+
+  void Stop() override {
+    PS_VLOG(1) << my_node_.ShortDebugString() << " is stopping";
+    Van::Stop();
+    should_stop_ = true;
+    polling_thread_->join();
+    polling_thread_.reset();
+
+    for (const auto& it : contexts_) {
+        it.second->Cleanup();
+    }
+
+    rx_pool_.reset();
+  }
+
+  int Bind(Node &node, int max_retry) override {
+    contexts_.reserve(my_node_.num_ports);
+
+    UCX_LOG(2, "Start/Bind UCX Van, num ports " << my_node_.num_ports);
+
+    for (int i = 0; i < my_node_.num_ports; ++i) {
+      int dev_id = my_node_.dev_ids[i];
+      CHECK_GE(dev_id, 0);
+
+      if (my_node_.dev_types[i] == GPU) {
+        cudaError_t cerr = cudaSetDevice(dev_id);
+
+        if (cudaSuccess != cerr) {
+          LOG(ERROR) << "Failed to set device " << dev_id << ": " << cerr;
         }
       }
+      contexts_[dev_id] = std::make_unique<UCXContext>(rx_pool_.get(), i);
+      contexts_[dev_id]->Init(&my_node_, this);
+      UCX_LOG(2, "Create ctx[" << i << "]: dev id " << dev_id << ", port "
+              << my_node_.ports[i]);
     }
 
-    char *buf;
-    size_t page_size = sysconf(_SC_PAGESIZE);
-    int ret          = posix_memalign((void**)&buf, page_size, size);
-    CHECK(!ret) << "posix_memalign error: " << strerror(ret);
-    CHECK(buf);
-    memset(buf, 0, size);
-    rpool_[key][node_id] = std::make_pair(buf, size);
+    polling_thread_.reset(new std::thread(&UCXVan::PollUCX, this));
 
-    return buf;
+    for (int i = 0; i < node.num_ports; ++i) {
+      contexts_[i]->Listen(node.ports[i]);
```

Comment:
not sure.. Do node.ports[] and node.dev_ids[] have the same indexation? (i. e. node.ports[i] corresponds to node.dev_ids[i])
Or node.ports[] are indexed by device ids? @eric-haibin-lin

Reply:
They have the same indexation.

Reply:
I just thought `contexts_[i]` on this line is inconsistent with line 794 through line 797.

Reply:
`contexts_` is a member of UCX van, I think better index them by dev_id, because when we do send we get a device_id, not its index in the node structure.

Reply:
suppose there's one port and one dev_id: `node.dev_ids[0] = 5` and `node.ports[0] = 12345`, line 803 will be using `contexts_[0]` that's not initialized.

Reply:
yes, you are right, will fix